### PR TITLE
feat: guide user for contract creation and add 4 contract templates

### PR DESCRIPTION
**Pull request by AlexD10S**
This PR includes an interactive user guide for contract generation and have an structure to manage different Contract template.

-  Interactive user guide for contract generation when no attributes has been input by the user: `pop new contract`:
Captura de pantalla 2024-06-10 a las 8 56 25

- Support diverse contract templates: 
              1. Standard: The default contract generate with `cargo-contract`
              2. ERC20: The implementation of the ERC-20 standard in Solidity using ink!
              3. ERC721: The implementation of the ERC-721 standard in Solidity using ink!
              4. ERC1155: The implementation of the ERC-1155 standard in Solidity using ink!
All templates were selected from: https://use.ink/examples/smart-contracts and the repository: https://github.com/use-ink/ink-examples/ 

**Considerations for reviewers**
- To fetch a contract template, the steps are as follows:
     1. Clone the [ink-examples](https://github.com/use-ink/ink-examples/ ) repository into a temporary folder.
     2. Copy only the necessary files from the selected template into the contract folder. 
    As indicated in the [code](https://github.com/r0gue-io/pop-cli/blob/alex/contracts-template/crates/pop-contracts/src/new.rs#L83), if future templates include directories, additional functionality will need to be added to support copying these directories.
- Renaming the template files to have the name chosen by the user. The same process that `cargo-contract` uses to rename the standard contract generation:
     1. Change the name in the Cargo.toml file, as shown [here](https://github.com/use-ink/cargo-contract/blob/master/crates/build/templates/new/_Cargo.toml#L2).
     2. Update the name in the code, as demonstrated [here](https://github.com/use-ink/cargo-contract/blob/master/crates/build/templates/new/lib.rs#L4).

- Since we are now generating contracts without using cargo-contract build function, I have added a check for contract naming with the `is_valid_contract_name` function. This function provides the same functionality as `cargo-contract`, as shown [here](https://github.com/use-ink/cargo-contract/blob/master/crates/build/src/new.rs#L39).

- This implementation introduces a basic common crate that includes all `Git` functionality and the `replace_in_file` function to avoid code duplication. In a future PR, we can refactor this crate to include more functions that are not specific to parachain or contracts. For now, it contains only the functionalities used in both crates. Closes https://github.com/r0gue-io/pop-cli/issues/124.

- Currently, the only provider is the official ink repository. If additional providers are added, such as for parachains, only a small amount of extra code will be required.

### CODE REVIEW

File: crates/pop-cli/src/commands/new/contract.rs [hunk 1]
```diff
@@ -1,74 +1,203 @@
...
+			Some(template) => template.clone(),
+			None => Template::Standard, // Default template
+		};
+
+		generate_contract_from_template(name, contract_config.path, &template)?;
 		Ok(())
 	}
 }
 
+async fn guide_user_to_generate_contract() -> Result<NewContractCommand> {
```
Comment (Daanvdplas) [inline crates/pop-cli/src/commands/new/contract.rs:59]: I don't understand all the newlines you are using

File: crates/pop-cli/src/commands/new/contract.rs [hunk 2]
```diff
@@ -1,74 +1,203 @@
...
+	fs::create_dir_all(contract_path.as_path())?;
+
+	let spinner = cliclack::spinner();
+	spinner.start("Generating contract...");
+
+	create_smart_contract(&name, contract_path.as_path(), template)?;
+
+	spinner.clear();
+
+	// replace spinner with success
```
Comment (Daanvdplas) [inline crates/pop-cli/src/commands/new/contract.rs:110]: Can we begin comments with a capital letter :)

File: crates/pop-cli/src/commands/new/contract.rs [hunk 3]
```diff
@@ -1,74 +1,203 @@
...
+	clear_screen()?;
+
+	Ok(NewContractCommand {
+		name: Some(name),
+		path: Some(PathBuf::from(path)),
+		template: Some(template.clone()),
+	})
+}
+
+fn generate_contract_from_template(
```
Comment (Daanvdplas) [inline crates/pop-cli/src/commands/new/contract.rs:89]: Unnecessary newlines again TMO

File: crates/pop-contracts/src/new.rs [hunk 4]
```diff
@@ -1,53 +1,130 @@
...
+			"Contract names must begin with an alphabetic character".to_owned(),
+		));
+	}
+	Ok(())
+}
+
+fn create_standard_contract(name: &str, canonicalized_path: PathBuf) -> Result<()> {
 	let parent_path = canonicalized_path
 		.parent()
 		// If the parent directory cannot be retrieved (e.g., if the path has no parent),
```
Comment (Daanvdplas) [inline crates/pop-contracts/src/new.rs:50]:
```suggestion
		// If the parent directory cannot be retrieved (e.g. if the path has no parent),
```

File: crates/pop-contracts/src/new.rs [hunk 5]
```diff
@@ -1,53 +1,130 @@
...
 		// return a NewContract variant indicating the failure.
 		.ok_or(Error::NewContract("Failed to get parent directory".to_string()))?;
-
-	// Create a new contract project with the provided name in the parent directory.
 	new_contract_project(&name, Some(parent_path))
 		// If an error occurs during the creation of the contract project,
 		// convert it into a NewContract variant with a formatted error message.
-		.map_err(|e| Error::NewContract(format!("{}", e)))
+		.map_err(|e| Error::NewContract(format!("{}", e)))?;
+	return Ok(());
```
Comment (Daanvdplas) [inline crates/pop-contracts/src/new.rs:57]:
```suggestion
	Ok(())
```

File: crates/pop-contracts/src/utils/helpers.rs [hunk 6]
```diff
@@ -28,6 +34,29 @@ pub fn parse_account(account: &str) -> Result<<DefaultConfig as Config>::Account
...
 
+pub fn canonicalized_path(target: &Path) -> Result<PathBuf, Error> {
+	// Canonicalize the target path to ensure consistency and resolve any symbolic links.
+	return target
+		.canonicalize()
+		// If an I/O error occurs during canonicalization, convert it into an Error enum variant.
+		.map_err(|e| Error::IO(e));
+}
+
+pub fn replace_in_file(file_path: PathBuf, replacements: HashMap<&str, &str>) -> Result<(), Error> {
```
Comment (Daanvdplas) [inline crates/pop-contracts/src/utils/helpers.rs:45]: missing docs, pub function

File: crates/pop-contracts/src/utils/helpers.rs [hunk 7]
```diff
@@ -28,6 +34,29 @@ pub fn parse_account(account: &str) -> Result<<DefaultConfig as Config>::Account
 		.map_err(|e| Error::AccountAddressParsing(format!("{}", e)))
 }
 
+pub fn canonicalized_path(target: &Path) -> Result<PathBuf, Error> {
+	// Canonicalize the target path to ensure consistency and resolve any symbolic links.
+	return target
```
Comment (Daanvdplas) [inline crates/pop-contracts/src/utils/helpers.rs:39]:
```suggestion
	target
```

File: crates/pop-contracts/src/utils/helpers.rs (hunk 6 above)
Comment (Daanvdplas) [reply crates/pop-contracts/src/utils/helpers.rs:45]: Untested

File: crates/pop-contracts/src/utils/helpers.rs [hunk 8]
```diff
@@ -28,6 +34,29 @@ pub fn parse_account(account: &str) -> Result<<DefaultConfig as Config>::Account
 		.map_err(|e| Error::AccountAddressParsing(format!("{}", e)))
 }
 
+pub fn canonicalized_path(target: &Path) -> Result<PathBuf, Error> {
```
Comment (Daanvdplas) [inline crates/pop-contracts/src/utils/helpers.rs:37]: Untested

File: crates/pop-contracts/src/utils/git.rs [hunk 9]
```diff
@@ -91,6 +144,28 @@ mod tests {
 			.await
 	}
 
+	#[test]
+	fn test_convert_to_ssh_url() {
```
Comment (Daanvdplas) [inline crates/pop-contracts/src/utils/git.rs:148]: 👌

File: crates/pop-contracts/src/utils/git.rs [hunk 10]
```diff
@@ -1,14 +1,62 @@
...
 use anyhow::Result;
+use git2::{build::RepoBuilder, FetchOptions, RemoteCallbacks, Repository};
+use git2_credentials::CredentialHandler;
+use std::path::Path;
 use url::Url;
 
-use crate::errors::Error;
+/// A helper for handling Git operations.
+pub struct Git;
+impl Git {
```
Comment (Daanvdplas) [inline crates/pop-contracts/src/utils/git.rs:11]: Can we test those functions?

File: crates/pop-contracts/src/templates.rs [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+		self.get_str("Repository").ok_or(Error::RepositoryMissing)
+	}
+	/// Get the list of templates supported.
+	pub fn templates() -> &'static [Template] {
+		Template::VARIANTS
+	}
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (Daanvdplas) [inline crates/pop-contracts/src/templates.rs:76]: 👌

**Review by Daanvdplas — COMMENTED** [review]
Nicely done! Left some nitpicks. 

Very curious how these templates will evolve!

File: crates/pop-contracts/src/templates.rs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+	EnumString,
+	Eq,
+	Hash,
+	PartialEq,
+	VariantArray,
+)]
+pub enum Template {
+	/// A minimalist contract template.
+	#[default]
+	#[strum(serialize = "standard", message = "Standard", detailed_message = "Our 'Hello World")]
```
[the remark this answers is not in the document]
Comment (AlexD10S) [reply crates/pop-contracts/src/templates.rs:26]: Is the description the Flipper Contract has in the ink docs: https://use.ink/examples/smart-contracts

File: crates/pop-contracts/src/templates.rs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+	ERC1155,
+}
+
+impl Template {
+	/// Get the template's name.
+	pub fn name(&self) -> &str {
+		self.get_message().unwrap_or_default()
+	}
+
+	/// Get the detailed message of the template.
```
[the remark this answers is not in the document]
Comment (AlexD10S) [reply crates/pop-contracts/src/templates.rs:60]: Let's keep consistency with parachains: https://github.com/r0gue-io/pop-cli/blob/main/crates/pop-parachains/src/templates.rs#L56

**Comment by peterwht** [comment]
Is it possible to introduce one more "directory" of contracts?

- What provider
    - ink! core
    - Pop Network
    - etc.
- What category:
    - ERC Standards
        - ERC20
        - ERC-721
   - PSP standards
   -

File: crates/pop-contracts/src/templates.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Daanvdplas) [reply crates/pop-contracts/src/templates.rs:60]: Should be description then TMO 🙈

**Comment by Daanvdplas** [comment]
> Is it possible to introduce one more "directory" of contracts?

Should it not be only PSP then?

File: crates/pop-contracts/src/new.rs [hunk 14]
```diff
@@ -1,53 +1,130 @@
...
+
+fn extract_contract_files(
+	contract_name: String,
+	repo_folder: &Path,
+	target_folder: &Path,
+) -> Result<()> {
+	let contract_folder = repo_folder.join(contract_name);
+	for entry in fs::read_dir(&contract_folder)? {
+		let entry = entry?;
+		// The currently available templates contain only files: lib.rs and Carg.toml. The `frontend` folder is being ignored.
```
Comment (peterwht) [inline crates/pop-contracts/src/new.rs:83]:
```suggestion
		// The currently available templates contain only files: lib.rs and Cargo.toml. The `frontend` folder is being ignored.
```

File: crates/pop-contracts/src/templates.rs [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+	EnumString,
+	Eq,
+	Hash,
+	PartialEq,
+	VariantArray,
+)]
+pub enum Template {
+	/// A minimalist contract template.
+	#[default]
+	#[strum(serialize = "standard", message = "Standard", detailed_message = "Our 'Hello World")]
```
Comment (peterwht) [inline crates/pop-contracts/src/templates.rs:26]: The name flipper is well known for those that use ink!. I think it should be mentioned somewhere here.
```suggestion
	#[strum(serialize = "standard", message = "Standard", detailed_message = "ink!'s 'Hello World': Flipper")]
```

File: crates/pop-contracts/src/templates.rs [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+		detailed_message = "The implementation of the ERC-721 standard in ink!",
+		props(Repository = "https://github.com/paritytech/ink-examples")
+	)]
+	ERC721,
+	/// The implementation of the ERC-1155 standard in ink!
+	#[strum(
+		serialize = "erc1155",
+		message = "Erc1155",
+		detailed_message = "The implementation of the ERC-1155 standard in ink!",
+		props(Repository = "https://github.com/paritytech/ink-examples")
```
Comment (peterwht) [inline crates/pop-contracts/src/templates.rs:49]: Please update all links to use https://github.com/use-ink/ink-examples

Also, (per my longer comment above), having an additional category would be nice to structure all ERCs together, all PSPs together, etc.

File: crates/pop-contracts/src/templates.rs [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+			("erc721".to_string(), Template::ERC721),
+			("erc1155".to_string(), Template::ERC1155),
+		])
+	}
+
+	fn templates_urls() -> HashMap<String, &'static str> {
+		HashMap::from([
+			("erc20".to_string(), "https://github.com/paritytech/ink-examples"),
+			("erc721".to_string(), "https://github.com/paritytech/ink-examples"),
+			("erc1155".to_string(), "https://github.com/paritytech/ink-examples"),
```
Comment (peterwht) [inline crates/pop-contracts/src/templates.rs:93]: links

**Review by peterwht — CHANGES_REQUESTED** [review]
I am really excited to see this merged in!! 

I'd encourage posting your progress of this on twitter :)

File: crates/pop-cli/src/commands/new/contract.rs [hunk 18]
```diff
@@ -1,74 +1,203 @@
...
+		long,
+		help = "Template to use.",
+		value_parser = crate::enum_variants!(Template)
+	)]
+	pub(crate) template: Option<Template>,
 }
 
 impl NewContractCommand {
-	pub(crate) async fn execute(&self) -> anyhow::Result<()> {
+	pub(crate) async fn execute(&self) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (AlexD10S) [reply crates/pop-cli/src/commands/new/contract.rs:32]: Refactored here: https://github.com/r0gue-io/pop-cli/pull/217

File: crates/pop-contracts/src/utils/helpers.rs (hunk 6 above)
Comment (AlexD10S) [reply crates/pop-contracts/src/utils/helpers.rs:60]: Actually, it did have coverage: https://app.codecov.io/gh/r0gue-io/pop-cli/blob/alex%2Fcontracts-template/crates%2Fpop-contracts%2Fsrc%2Futils%2Fhelpers.rs#L45

But I agree that it's better to have a specific test for this function. I've added it.

File: crates/pop-contracts/src/utils/helpers.rs (hunk 8 above)
Comment (AlexD10S) [reply crates/pop-contracts/src/utils/helpers.rs:44]: Same as above.

File: crates/pop-contracts/src/templates.rs [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+	Hash,
+	PartialEq,
+	VariantArray,
+)]
+pub enum Template {
+	/// A minimalist contract template.
+	#[default]
+	#[strum(serialize = "standard", message = "Standard", detailed_message = "Our 'Hello World")]
+	Standard,
+	/// The implementation of the ERC-20 standard in Solidity using ink!
```
[the remark this answers is not in the document]
Comment (AlexD10S) [reply crates/pop-contracts/src/templates.rs]: Continue the discussion here https://github.com/r0gue-io/pop-cli/pull/201#issuecomment-2183035580

File: README.md [hunk 20]
```diff
@@ -124,13 +124,31 @@ pop up parachain -f ./tests/zombienet.toml -p https://github.com/r0gue-io/pop-no
...
 
 ```sh
 # Create a minimal Smart Contract
 pop new contract my_contract
 ```
 
+`pop-cli` supports diverse project templates, to use a specific one use the flag `--template`:
+
+```sh
+# Create an ERC-20 standard in Solidity using ink!
```
[the remark this answers is not in the document]
Comment (AlexD10S) [reply README.md]: Let's discuss here: https://github.com/r0gue-io/pop-cli/pull/201#issuecomment-2183035580

**Comment by AlexD10S** [comment]
> Is it possible to introduce one more "directory" of contracts?
> 
> * What provider
>   
>   * ink! core
>   * Pop Network
>   * etc.
>   * What category:
>     
>     * ERC Standards
>       
>       * ERC20
>       * ERC-721
>     * PSP standards

Yes, just like we do for parachains, we can have different Providers for templates. This PR introduces some basic templates that are a bit more complex than Flipper. But we can also add many more templates right from the start.

File: crates/pop-contracts/src/new.rs [hunk 21]
```diff
@@ -1,65 +1,247 @@
...
+		let temp_dir = tempfile::tempdir()?;
+		let config = temp_dir.path().join("Cargo.toml");
+		let mut config_file = fs::File::create(config.clone())?;
+		writeln!(
+			config_file,
+			r#"
+				[package]
+				name = "erc20"
+				version = "5.0.0"
+				authors = ["R0gue"]
```
Comment (al3mart) [inline crates/pop-contracts/src/new.rs:218]:
```suggestion
				authors = ["R0GUE"]
```

**Review by al3mart — CHANGES_REQUESTED** [review]
I think I agree with Peter in the sense that a directory like set of providers might work better when it comes to extend and integrate providers in the future.

My suggestion would be

> Choose provider
- use-ink!
- - ERC-20
- - ERC-721
- - ERC-1155
- - PSP-22
- - PSP-34
- Pop Network
- - Pop API template
- Others...
- - Other's catalog...

**Review by peterwht — APPROVED** [review]

**Comment by AlexD10S** [comment]
> > Is it possible to introduce one more "directory" of contracts?
> 
> Should it not be only PSP then?

After discussing with Peter, we've decided to only include templates from the ink! repository for now, which do not include any PSP implementation.
Following Peter's PR merged here: https://github.com/r0gue-io/pop-cli/pull/220 we've reorganized the templates into different categories, starting with  `examples` and `erc`.

File: crates/pop-contracts/src/new.rs [hunk 22]
```diff
@@ -1,65 +1,243 @@
...
+
+pub fn is_valid_contract_name(name: &str) -> Result<(), Error> {
+	if !name.chars().all(|c| c.is_alphanumeric() || c == '_') {
+		return Err(Error::InvalidName(
+			"Contract names can only contain alphanumeric characters and underscores".to_owned(),
+		));
+	}
+	if !name.chars().next().map(|c| c.is_alphabetic()).unwrap_or(false) {
+		return Err(Error::InvalidName(
+			"Contract names must begin with an alphabetic character".to_owned(),
```
Comment (brunopgalvao) [inline crates/pop-contracts/src/new.rs:37]:
```suggestion
			"Contract names must begin with an alphabetical character".to_owned(),
```

File: crates/pop-contracts/src/new.rs [hunk 23]
```diff
@@ -1,65 +1,243 @@
...
 }
 
 #[cfg(test)]
 mod tests {
 	use super::*;
 	use anyhow::{Error, Result};
-	use std::fs;
-	use tempfile;
+	use std::{fs, io::Write};
+	use tempfile::{self};
```
Comment (brunopgalvao) [inline crates/pop-contracts/src/new.rs:110]: I have never seen importing `self`.

File: crates/pop-contracts/src/templates.rs [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+	EnumMessage,
+	EnumProperty,
+	EnumString,
+	Eq,
+	Hash,
+	PartialEq,
+	VariantArray,
+)]
+pub enum Template {
+	/// A minimalist contract template.
```
Comment (brunopgalvao) [inline crates/pop-contracts/src/templates.rs:79]:
```suggestion
	/// A minimal contract template.
```

File: crates/pop-contracts/src/templates.rs [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+		serialize = "standard",
+		message = "Standard",
+		detailed_message = "ink!'s 'Hello World': Flipper",
+		props(ContractType = "Examples")
+	)]
+	Standard,
+	/// The implementation of the ERC-20 standard in ink!
+	#[strum(
+		serialize = "erc20",
+		message = "Erc20",
```
Comment (brunopgalvao) [inline crates/pop-contracts/src/templates.rs:91]:
```suggestion
		message = "ERC-20",
```

File: crates/pop-contracts/src/templates.rs [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+		detailed_message = "The implementation of the ERC-20 standard in ink!",
+		props(ContractType = "ERC", Repository = "https://github.com/use-ink/ink-examples")
+	)]
+	ERC20,
+	/// The implementation of the ERC-721 standard in ink!
+	#[strum(
+		serialize = "erc721",
+		message = "Erc721",
```
Comment (brunopgalvao) [inline crates/pop-contracts/src/templates.rs:99]:
```suggestion
		message = "ERC-721",
```

File: crates/pop-contracts/src/templates.rs [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+		detailed_message = "The implementation of the ERC-721 standard in ink!",
+		props(ContractType = "ERC", Repository = "https://github.com/use-ink/ink-examples")
+	)]
+	ERC721,
+	/// The implementation of the ERC-1155 standard in ink!
+	#[strum(
+		serialize = "erc1155",
+		message = "Erc1155",
```
Comment (brunopgalvao) [inline crates/pop-contracts/src/templates.rs:107]:
```suggestion
		message = "ERC-1155",
```

**Review by brunopgalvao — APPROVED** [review]

File: crates/pop-contracts/src/new.rs (hunk 23 above)
Comment (AlexD10S) [reply crates/pop-contracts/src/new.rs:110]: True. No needed.

File: crates/pop-contracts/src/templates.rs (hunk 25 above)
Comment (AlexD10S) [reply crates/pop-contracts/src/templates.rs:91]: Got the name of the contract from here: https://github.com/use-ink/ink-examples/blob/main/erc20/lib.rs#L10 And used this message to replace the name in the contract once cloned.

File: crates/pop-contracts/src/templates.rs (hunk 26 above)
Comment (AlexD10S) [reply crates/pop-contracts/src/templates.rs:99]: Same as above.

File: crates/pop-contracts/src/templates.rs (hunk 27 above)
Comment (AlexD10S) [reply crates/pop-contracts/src/templates.rs:107]: Same as above.

**Review by al3mart — APPROVED** [review]
Looking good.

I'm fine with whatever the outcome is regarding the open comments about the name of the examples.

So proceeding with an approve.

**Review by peterwht — APPROVED** [review]

**Pull request merged**